puzzle-utils/words/ladder.py

```diff
@@ -0,0 +1,92 @@
+from .search import search
+
+def ladder(word):
+    """
+    Given a word, find all "adjacent" words in the dictionary
+    i.e. all words with edit distance of 1 from the given word
+    """
+    found_words = []
+    for i in range(len(word)):
+        pattern = list(word)
+        pattern[i] = '.'
+        search_results = search("^" + "".join(pattern) + "$")
+        for result in search_results:
+            if result != word and result not in found_words:
+                found_words.append(result)
+    return found_words
+
+def edit_distance(left_word, right_word):
+    """
+    This assumes the lengths of the words are the same.
+    The result is the edit distance of the two words (number of
+    substitutions required to tranform left_word into right_word).
+    """
+    distance = 0;
+    for i in range(len(left_word)):
+        if left_word[i] != right_word[i]:
+            distance += 1
+    return distance
+
+def follow_ladder_path(word_path, right_word, cached_searches, depth, max_depth):
+    """
+    Helper to recursively find a word ladder path
+    """
+    # We're too deep, this will take too long
+    if depth > max_depth:
+        return None
+
+    # Find words closer to destination (avoiding cycles)
```

Comment:
I can see there being situations where the puzzle may require more steps then what is possible but that is probably a massive scope creep of what was intended.

Reply:
I'll make this a param